src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -188,10 +189,49 @@ public Stream<MemorySegment> elements(MemoryLayout elementLayout) {
         return StreamSupport.stream(spliterator(elementLayout), false);
     }
 
+    // FILL_NATIVE_THRESHOLD must be a power of two and should be greater than 2^3
+    // Update the value for Aarch64 once 8338975 is fixed.
+    private static final long FILL_NATIVE_THRESHOLD = 1L << (Architecture.isAARCH64() ? 10 : 5);
+
     @Override
-    public final MemorySegment fill(byte value){
-        checkAccess(0, length, false);
-        SCOPED_MEMORY_ACCESS.setMemory(sessionImpl(), unsafeGetBase(), unsafeGetOffset(), length, value);
+    public final MemorySegment fill(byte value) {
+        checkReadOnly(false);
+        // 0...0X...XXXX implies: 0 <= length < FILL_NATIVE_LIMIT
+        if ((length & -FILL_NATIVE_THRESHOLD) == 0) {
+            // Handle smaller segments directly without transitioning to native code
+            if (length == 0) {
+                // Implicit state check
+                checkValidState();
+                return this;
+            }
+            final long u = Byte.toUnsignedLong(value);
+            final long longValue = u << 56 | u << 48 | u << 40 | u << 32 | u << 24 | u << 16 | u << 8 | u;
```

Comment:
this can be u * 0xFFFFFFFFFFFFL if value != 0 and just 0L if not: not sure if fast(er), need to measure.

Most of the time filling is happy with 0 since zeroing is the most common case

Reply:
> this can be u * 0xFFFFFFFFFFFFL if value != 0 and just 0L if not: not sure if fast(er), need to measure.
> 
> Most of the time filling is happy with 0 since zeroing is the most common case

It's a clever trick. However, I was looking at similar tricks and found that the time spent here is irrelevant (e.g. I tried to always force `0` as the value, and couldn't see any difference).

Reply:
If I run:

```
    @Benchmark
    public long shift() {
        return ELEM_SIZE << 56 | ELEM_SIZE << 48 | ELEM_SIZE << 40 | ELEM_SIZE << 32 | ELEM_SIZE << 24 | ELEM_SIZE << 16 | ELEM_SIZE << 8 | ELEM_SIZE;
    }

    @Benchmark
    public long mul() {
        return ELEM_SIZE * 0xFFFF_FFFF_FFFFL;
    }
```
Then I get:
```
Benchmark       (ELEM_SIZE)  Mode  Cnt  Score   Error  Units
TestFill.mul             31  avgt   30  0.586 ? 0.045  ns/op
TestFill.shift           31  avgt   30  0.938 ? 0.017  ns/op
```
On my M1 machine.

Reply:
I found similar small improvements to be had (I wrote about them offline) when replacing the bitwise-based tests (e.g. `foo & 4 != 0`) with a more explicit check for `remainingBytes >=4`. Seems like bitwise operations are not as optimized (or perhaps the assembly instructions for them is overall more convoluted - I haven't checked).

Reply:
I've tried 

```
final long longValue = Byte.toUnsignedLong(value) * 0x0101010101010101L;
```

But it had the same performance as explicit bit shifting on M1.

Reply:
@minborg the ` ELEM_SIZE` is a `Param` field right? Just to be 100% sure of it...

Reply:
yes it is. But I think the reason we are not seeing any difference is that in the fill benchmarks, we are always using a value of zero. We should keep this trick in mind for later...